### PR TITLE
Codec Guide

### PR DESCRIPTION
**Pull request by enjarai**
This PR adds the codec guide I've previously written for https://fabric.moddedmc.wiki/, with some slight updates, fixes and tweaks. As I'm not exactly sure how this wiki will be structured, I'd like to leave integrating it into navigation and the rest of the wiki to the main contributors.

If you have any questions or would like changes made, please let me know.

### CODE REVIEW

**Review by IMB11 — CHANGES_REQUESTED** [review]
Please put any code snippets into the reference mod (see contributing style guidelines). That's the only issue so far! Looking good!

**Comment by enjarai** [comment]
Due to how the code snippets are structured, I'm not sure how feasible it is to integrate them into an actual mod without significant restructuring. Most of the code, while valid, is very much simplified and compacted to address more complex topics.

**Comment by IMB11** [comment]
You can manipulate the way they are displayed using the vitepress snippet component options - it doesn't necessarily mean you have to restructure your codeblocks

File: develop/codecs.md [hunk 1]
```diff
@@ -0,0 +1,440 @@
...
+    // ...
+}
+
+// Create a registry to map identifiers to bean types
+Registry<BeanType<?>> beanTypeRegistry = new SimpleRegistry<>(new Identifier("example", "bean_types"), Lifecycle.stable());
+
+// Register the bean types, dont forget to store these in a 
+// static constant and return them for `getType` in their respective subclasses.
+beanTypeRegistry.register(new Identifier("example", "stringy_bean"), new BeanType<>(StringyBean.CODEC));
+beanTypeRegistry.register(new Identifier("example", "counting_bean"), new BeanType<>(CountingBean.CODEC));
```
Comment (deirn) [inline develop/codecs.md]: The `BeanType` should be initialized as a static field somewhere the `Bean` implementation can access.

Also, I don't think there's a method directly in the `Registry` instance to register, only the static method.

File: develop/codecs.md [hunk 2]
```diff
@@ -0,0 +1,440 @@
...
+        Codec.INT.fieldOf("counting_number").forGetter(CountingBean::getCountingNumber)
+    ).apply(instance, CountingBean::new));
+
+    public final int countingNumber;
+
+    // ...
+}
+
+// Create a registry to map identifiers to bean types
+Registry<BeanType<?>> beanTypeRegistry = new SimpleRegistry<>(new Identifier("example", "bean_types"), Lifecycle.stable());
```
Comment (deirn) [inline develop/codecs.md]: `SimpleRegistry` constructor doesn't accept an `Identifier` but `RegistryKey` instead.

File: develop/codecs.md [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+[converted `MapCodec`s](#mapcodec-not-to-be-confused-with-codecltmapgt) or 
+[record codecs](#merging-codecs-for-record-like-classes). 
+The resulting codec will serialize to a map combining the fields of both codecs used.
+
+For example, running this code:
+
+```java
+// Create two seperate boxed codecs
+MapCodec<Integer> firstCodec = Codec.INT.fieldOf("i_am_number").codec();
+MapCodec<Boolean> secondCodec = Codec.BOOL.fieldOf("this_statement_is_false").codec();
```
Comment (deirn) [inline develop/codecs.md]: `MapCodec#codec` returns a `Codec` instance.

```suggestion
Codec<Integer> firstCodec = Codec.INT.fieldOf("i_am_number").codec();
Codec<Boolean> secondCodec = Codec.BOOL.fieldOf("this_statement_is_false").codec();
```

File: develop/codecs.md [hunk 4]
```diff
@@ -0,0 +1,440 @@
...
+```java
+// Without a default value
+MapCodec<Optional<BlockPos>> optionalCodec = BlockPos.CODEC.optionalFieldOf("pos");
+
+// With a default value
+MapCodec<BlockPos> optionalCodec = BlockPos.CODEC.optionalFieldOf("pos", BlockPos.ORIGIN);
+```
+
+Do note that optional fields will silently ignore any errors that may occur during deserialization. This means that if
+the field is present, but the value is invalid, the field will always be deserialized as the default value.
```
Comment (deirn) [inline develop/codecs.md:199]: Minecraft (not DFU!) provides a method that creates an optional codec that errors if it failed to deserialize. Might worth mentioning it here.

[`Codecs#createStrictOptionalFieldCodec`](https://maven.fabricmc.net/docs/yarn-1.20.4+build.3/net/minecraft/util/dynamic/Codecs.html#createStrictOptionalFieldCodec(com.mojang.serialization.Codec,java.lang.String))

**Review by deirn — CHANGES_REQUESTED** [review]

**Comment by IMB11** [comment]
@enjarai You should be able to use the snippet syntax from #19 now. (See discord thread)

File: develop/codecs.md [hunk 5]
```diff
@@ -0,0 +1,440 @@
+---
+title: Codecs
+description: A comprehensive guide for understanding and using Mojang's codec system for serializing and deserializing objects.
```
Comment (IMB11) [inline develop/codecs.md:3]:
```suggestion
description: A comprehensive guide for understanding and using Mojang's codec system for serializing and deserializing objects.
authors:
	- enjarai
```

Comment (IMB11) [reply develop/codecs.md:3]:
```suggestion
description: A comprehensive guide for understanding and using Mojang's codec system for serializing and deserializing objects.
authors:
  - enjarai
```

File: develop/codecs.md [hunk 6]
```diff
@@ -0,0 +1,442 @@
...
+authors:
+	- enjarai
```
Comment (IMB11) [inline develop/codecs.md]:
```suggestion
  - enjarai
```

**Comment by enjarai** [comment]
@deirn thanks for the suggestions, I've incorporated them into the guide.

**Review by IMB11 — DISMISSED** [review]
Seems good! :+1:

**Review by IMB11 — DISMISSED** [review]

**Review by IMB11 — APPROVED** [review]

**Comment by IMB11** [comment]
LGTM

**Comment by enjarai** [comment]
If checks fail again I will cry.

**Pull request merged**